File: IDATT2106 Systemutvikling 2 med smidig prosjekt/backend/src/main/java/com/ntnu/gidd/controller/AuthenticationController.java
package com.ntnu.gidd.controller;

import com.ntnu.gidd.dto.JwtTokenResponse;
import com.ntnu.gidd.dto.User.UserPasswordForgotDto;
import com.ntnu.gidd.dto.User.UserPasswordUpdateDto;
import com.ntnu.gidd.exception.*;
import com.ntnu.gidd.model.PasswordResetToken;
import com.ntnu.gidd.security.config.JWTConfig;
import com.ntnu.gidd.security.service.JwtService;
import com.ntnu.gidd.service.User.UserService;
import com.ntnu.gidd.util.Response;
import com.ntnu.gidd.dto.User.UserPasswordResetDto;
import lombok.AllArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import javax.mail.MessagingException;
import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;
import java.security.Principal;
import java.util.UUID;

@Slf4j
@RestController
@RequestMapping("auth/")
@AllArgsConstructor
public class AuthenticationController {
	
	private JWTConfig jwtConfig;
	private JwtService jwtService;
	
	@Autowired
	private UserService userService;
	
	@GetMapping("/refresh-token/")
	public JwtTokenResponse refreshToken(HttpServletRequest request) {
		String header = request.getHeader(jwtConfig.getHeader());
		return jwtService.refreshToken(header);
	}
	
	@PostMapping("/change-password/")
	@ResponseStatus(HttpStatus.OK)
	public Response updatePassword(Principal principal, @RequestBody @Valid UserPasswordUpdateDto user) {
		userService.changePassword(principal, user);
		return new Response("Password was successfully changed");
	}
	
	/**
	 * Creates a reset password token and sends email if correct email i provided
	 *
	 * @param UserPasswordForgotDto of the users which should have its password reset
	 * @return
	 */
	@SneakyThrows
	@PostMapping("/forgot-password/")
	@ResponseStatus(HttpStatus.OK)
	public Response forgotPassword(@RequestBody UserPasswordForgotDto UserPasswordForgotDto) {
		userService.forgotPassword(UserPasswordForgotDto.getEmail());

		log.info("Email sent to {} for resetting password", UserPasswordForgotDto.getEmail());
		return new Response("An email for resetting password has been sent!");
	}
	
/**
* Reset the given users password, if a valid token is provided. 
*
* @param userPasswordResetDto userPasswordResetDto DTO the new passord the user want to change, the email of the user and the PasswordResetToken.
* @throws ResponseStatusException if the token or linked email is invalid.
**/
	@PostMapping("/reset-password/{passwordResetTokenId}/")
	@ResponseStatus(HttpStatus.OK)
	public Response resetPassword(@RequestBody @Valid UserPasswordResetDto userPasswordResetDto, @PathVariable UUID passwordResetTokenId) {
		userService.validateResetPassword(userPasswordResetDto, passwordResetTokenId);
		log.info("Password was changed for user {}", userPasswordResetDto.getEmail());
		return new Response("Password was changed successfully");
	}
	
}

File: IDATT2105 Full-stack applikasjonsutvikling/Rombestilling/backend/src/main/kotlin/ntnu/idatt2105/security/controller/AuthenticationController.kt
package ntnu.idatt2105.security.controller

import io.swagger.annotations.Api
import io.swagger.annotations.ApiParam
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.responses.ApiResponse
import ntnu.idatt2105.core.response.Response
import ntnu.idatt2105.security.dto.*
import ntnu.idatt2105.user.dto.DetailedUserDto
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.bind.annotation.PostMapping
import java.util.*
import javax.servlet.http.HttpServletRequest

@Api(value = "Authentication services", tags = ["Authentication Services"], description = "Authentication Services")
@RequestMapping("auth/")
interface AuthenticationController {

    @Operation(summary = "Log into the application", responses = [
        ApiResponse(responseCode = "200", description = "Success: Access and refresh token generated")])
    @PostMapping("/login")
    fun fakeLogin(
        @ApiParam("Login credentials") @RequestBody credentials: LoginRequest?,
    ) {
        throw IllegalStateException("This method shouldn't be called. It's implemented by Spring Security filters.")
    }

    @Operation(summary = "Refresh an old access token", responses = [
        ApiResponse(responseCode = "200", description = "Success"),
        ApiResponse(responseCode = "400", description = "Bad request: a new refresh token could not be issued"),
    ])
    @GetMapping("/refresh-token/")
    fun refreshToken(request: HttpServletRequest): JwtTokenResponse?

    @Operation(summary = "Send an email with a link to reset password", responses = [
        ApiResponse(responseCode = "201", description = "Created: reset password email was sent"),
        ApiResponse(responseCode = "400", description = "Bad request: reset password email was not sent"),
    ])
    @PostMapping("/forgot-password/")
    fun forgotPassword(@RequestBody email: ForgotPassword): ResponseEntity<Response>

    @Operation(summary = "Reset a password", responses = [
        ApiResponse(responseCode = "201", description = "Created: password was reset"),
        ApiResponse(responseCode = "400", description = "Bad request: password was not reset"),
    ])
    @PostMapping("/reset-password/{passwordResetTokenId}/")
    fun resetPassword(@PathVariable passwordResetTokenId: UUID, @RequestBody reset: ResetPasswordDto): ResponseEntity<Response>

    @Operation(summary = "Make a user admin", responses = [
        ApiResponse(responseCode = "200", description = "Success: user was made an admin"),
        ApiResponse(responseCode = "404", description = "Not found: user was not found"),
    ])
    @PostMapping("/make-admin/")
    fun makeAdmin(@RequestBody user: MakeAdminDto): ResponseEntity<DetailedUserDto>
}
